src/ILCompiler.Build.Tasks/src/ComputeManagedAssemblies.cs

```diff
@@ -0,0 +1,142 @@
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Reflection.Metadata;
+using System.Reflection.PortableExecutable;
+
+
+
+namespace Build.Tasks
+{
+    public class ComputeManagedAssemblies : Task
+    {
+        [Required]
+        public ITaskItem[] Assemblies
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The CoreRT-specific System.Private.* assemblies that must be used instead of the netcoreapp2.0 versions.
+        /// </summary>
+        [Required]
+        public ITaskItem[] SdkAssemblies
+        {
+            get;
+            set;
+        }
+
+        /// <summary>
+        /// The set of AOT-specific framework assemblies we currently need to use which will replace the same-named ones
+        /// in the app's closure.
+        /// </summary>
+        [Required]
+        public ITaskItem[] FrameworkAssemblies
+        {
+            get;
+            set;
+        }
+
+        [Output]
+        public ITaskItem[] ManagedAssemblies
+        {
+            get;
+            set;
+        }
+
+        [Output]
+        public ITaskItem[] AssembliesToSkipPublish
+        {
+            get;
+            set;
+        }
+
+        public override bool Execute()
+        {
+            var list = new List<ITaskItem>();
+            var assembliesToSkipPublish = new List<ITaskItem>();
+
+            ITaskItem[] assemblies = this.Assemblies;
+
+            var coreRTFrameworkAssembliesToUse = new HashSet<string>();
+
+            foreach (var x in SdkAssemblies)
+            {
+                coreRTFrameworkAssembliesToUse.Add(Path.GetFileName(x.ItemSpec));
+            }
+
+            foreach (var x in FrameworkAssemblies)
+            {
+                coreRTFrameworkAssembliesToUse.Add(Path.GetFileName(x.ItemSpec));
+            }
+
+            for (int i = 0; i < assemblies.Length; i++)
+            {
+                ITaskItem taskItem = assemblies[i];
+
+                // Skip crossgen images
+                if (taskItem.ItemSpec.EndsWith(".ni.dll", StringComparison.OrdinalIgnoreCase))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                // Skip the native apphost (whose name ends up colliding with the CoreRT output binary) and supporting assemblies
+                if (taskItem.ItemSpec.EndsWith("apphost.exe", StringComparison.OrdinalIgnoreCase) || taskItem.ItemSpec.Contains("hostfxr") || taskItem.ItemSpec.Contains("hostpolicy"))
+                {
+                    assembliesToSkipPublish.Add(taskItem);
+                    continue;
+                }
+
+                // Prototype aid - remove the native CoreCLR runtime pieces from the publish folder
```

Comment:
Same question - would it be skipped more naturally in the `foreach`?

Reply:
Yes, however CoreCLR assembly modules don't seem to have metadata attached, which to inspect. Hopefully someone more knowledgeable will correct me here.